### ISSUE TITLE
Imperative Migrations

### ISSUE THREAD

**Issue by flybayer** [prisma/migrate#492]
## Problem

Not all migrations can be declarative. For example, we need to add a custom index for performance reasons. We also need this as an escape hatch for anything prisma schema doesn't support. Lastly, we also need this for data migrations.

This is very important for any serious apps using prisma migrate.

## Suggested solution

A new command: `prisma migrate create`

This will create the necessary migration files. Similar as `prisma migrate save` but the files are just stubbed out.

Then the user will open the stubbed migration file and add their custom migration code.

Perhaps the imperative migration file can look something like this:

```ts
import {migrate} from '@prisma/migrate`

migrate({
  async up(prisma) {
    // SQL migration
    await prisma.executeRaw(`some SQL stuff`)

    // Data migration
    await prisma.someModel.create(/*..*/)
  },

  async down(prisma) {
    await prisma.someModel.delete(/*..*/)

    await prisma.executeRaw(`undo some SQL stuff`)
  }
})
```

**Comment by flybayer** [comment]
cc @peterp you'll probably be interested in this for Redwood too?

**Comment by karanpratapsingh** [comment]
@flybayer +1 really like this idea

**Comment by ryanking1809** [comment]
100% - This is the only thing stopping me from being able to fully utilise migrate.

I even think `steps.json` could even be added to this file and things could work more similarly to `sqlAlchemy` or Rails' `Active Record`

```js
migrate({
  async up(prisma) {
    // Prisma migration steps for intial setup
    // a lot of the time we need to migrate the data before we can drop columns etc
    // or setup defaults for non-nullable fields
    await prisma.db.addColumn(...)

    // Data migration
    await prisma.someModel.update(/*..*/)

    // Prisma migration for db cleanup
    // only once we know the data is safe in another location and defaults setup
    // can we drop columns and make columns non-nullable
    await prisma.db.alterColumn(...)
    await prisma.db.dropColumn(...)

    // Raw when needed
    await prisma.executeRaw(`some SQL stuff`)
  },

  async down(prisma) {
    await prisma.db.addColumn(...)
    //...
    await prisma.db.dropColumn(...)
  }
})
```

You would need to verify that the resulting db will match the schema but it would allow for much more complex migrations.

**Comment by flybayer** [comment]
@ryanking1809 yes!! I definitely thought about that — that would be super cool.

That also opens up the possibility to have a single `change()` hook that automatically generates up and down migrations, same as ActiveRecord.

```tsx
migrate({
  // Automatically generates both `up` and `down` migrations
  async change(prisma) {
    await prisma.db.addColumn(...)
    await prisma.db.addIndex(...)
  }
})
```

**Comment by Jolg42** [comment]
Thanks for the suggestion! We're currently working on Migrate so any feedback is welcome 😄 

@flybayer So what I see is the main driver for imperative migrations is to provide an escape hatch for unsupported features and data migrations. We really care about these two cases and want them covered.

If this was covered not by imperative migrations but before / after hooks will this work for you?
For example: before.ts / after.ts or even before.sql / after.sql called before or after a migration.

Let me know what you think

**Comment by flybayer** [comment]
I don't think that would be sufficient. Let's say I want to create an SQL function to use for default values. This is an imperative migration that's not tied to anything else. 

There's no existing migration for which I can add a before or after hook since this is a standalone thing.

Also for complex migrations requiring data migrations, only before after hooks would be very cumbersome. Example:

1) Make first change in schema file
2) Save that as a migration
3) Write the after hook for that
4) Make another change to schema file
5) Save that as a migration
6) Write the after hook for that second migration

**Comment by ryanking1809** [comment]
Yes, to extend what @flybayer is saying with some examples, say I had a `project` with `tasks`, the initially I might have done this in a `json` column because I'm still designing what tasks are.
```js
model Project {
  id   Int @id
  tasks Json 
}
```
After a bit of use, my customers want more features like task `completion`, and `dueDates`, and I decide to move tasks to it's own model. Ideally, I'd simply do this in one migration.
```js
model Project {
  id   Int @id
  tasks Task[]
}
model Task {
  id   Int @id
  project Project 
  completed   Boolean
  dueDate DateTime
}
```
And in the migration file I would be able to:

1. Create the task table
2. Migrate the task data from the project model to the task model
3. Drop the task column from projects

---

A before and after hook would force this to be executed in 2 migrations. And there's a little confusing on how to define the relationships in the `prisma.schema` file.

Migration Schema 1
```js
model Project {
  id   Int @id
  // this has to stay intact so we can migrate the data
  tasks Json 
}
model Task {
  id   Int @id
  // project relation not possible due to the definition being on one side
  // but the foreign key is required in the data so we just use the projectId Column for now
  projectId Int
  completed   Boolean
  dueDate DateTime
}
```

We can then migrate the data with and `after` hook, and only after that we can make the schema look they way we want.
```js
model Project {
  id   Int @id
  tasks Task[]
}
model Task {
  id   Int @id
  project Project 
  completed   Boolean
  dueDate DateTime
}
```

It works but is a lot more cumbersome. 

---

I've also had situations in the past where I've just been required to migrate data and make no changes to the database. An example maybe we previously decided to store the `length` of a task in `integer days` but now need more detail so the length will now be `integer minutes`. The column `length` remains an integer but we need to multiply out the data.

Whilst the data can migrated with a script. It's very useful to have it integrated into your deploy process. Especially if you're dealing with multiple databases. So being able to have a migration that isn't directly tied the the `prisma.schema` file would be amazing.

**Comment by davidmcnamee** [comment]
I agree with the feature -- lack of data migrations is the main thing holding me back from using prisma in all my projects. But I disagree with the suggested solution in terms of API.

For example, the suggested solution would produce code like this:

```typescript
// migration.ts

import { migrate } from '@prisma/migrate`

migrate({
  async up(prisma) {

    // update some fields
    await prisma.updateField({ tag: "UpdateField", model: "Following", field: "id", type: "Int" })

    // create some new directives, arguments, feilds, or models
    await prisma.createDirective({ tag: "CreateDirective", location: { path: { tag: "Field", model: "Following", field: "id" }, directive: "id" } })
    await prisma.createDirective({ tag: "CreateDirective", location: { path: { tag: "Field", model: "Following", field: "id" }, directive: "default" } })
    await prisma.createArgument({ tag: "CreateArgument", location: { tag: "Directive", path: { tag: "Field", model: "Following", field: "id" }, directive: "default" }, argument: "", value: "autoincrement()" })

    // Data migration
    await prisma.someModel.rows.forEach(r => {
      // move data around with typescript
    });
    await prisma.executeRaw(`some SQL stuff`) // ...or just use SQL

    // delete the old directives
    await prisma.deleteDirective({ tag: "DeleteDirective", location: { path: { tag: "Field", model: "Following", field: "id" }, directive: "unique" } })
  },

  async down(prisma) {
    // repeat EVERYTHING from above, but in the reverse order
  }
})
```

This seems like way too much generated/tedious code in my opinion, especially once you scale up. I think it would be a good idea to leverage the existing `steps.json` and simply insert a **new** type of step that has some tag similar to `ExternalMigration` that comes with a path to the migration file. Since most migrations can be boiled down to **`Create new columns/tables -> Move the data around -> Update columns/tables -> Drop columns/tables`**, this would allow you to seamlessly insert a "Data Migration" step somewhere in your generated migration. Here's an example of what I mean:

`migrations/20200717142452-my-migration/steps.json`
```json
{
  "version": "0.3.14-fixed",
  "steps": [
    {
      "tag": "UpdateField",
      "model": "Following",
      "field": "id",
      "type": "Int"
    },
    {
      "tag": "CreateDirective",
      "location": {
        "path": {
          "tag": "Field",
          "model": "Following",
          "field": "id"
        },
        "directive": "id"
      }
    },
    {
      "tag": "CreateDirective",
      "location": {
        "path": {
          "tag": "Field",
          "model": "Following",
          "field": "id"
        },
        "directive": "default"
      }
    },
    {
      "tag": "CreateArgument",
      "location": {
        "tag": "Directive",
        "path": {
          "tag": "Field",
          "model": "Following",
          "field": "id"
        },
        "directive": "default"
      },
      "argument": "",
      "value": "autoincrement()"
    },
    {
      "tag": "ExternalMigration", <--------- there it is!!!
      "path": "./data-migration.ts"
    },
    {
      "tag": "DeleteDirective",
      "location": {
        "path": {
          "tag": "Field",
          "model": "Following",
          "field": "id"
        },
        "directive": "unique"
      }
    }
  ]
}
```

```typescript
// migrations/20200717142452-my-migration/data-migration.ts
import { migrate } from '@prisma/migrate`

migrate({
  async up(prisma) {
    await prisma.someModel.rows.forEach(r => {
      // move data around with typescript
    });
    await prisma.executeRaw(`some SQL stuff`) // ...or just use SQL
  },

  async down(prisma) {
    await prisma.executeRaw('reverse the previous changes')
  }
})
```

It could even come with a json stub like `{"tag": "ExternalMigration", ...}` so that you can then directly copy/paste it into `steps.json` at whichever point you want it to execute. But primarily, I think it would be great if prisma could have this kind of intermediary data-migration step [rather than the `before.sh` and `after.sh` that are in the spec](https://github.com/prisma/specs/tree/master/migrate#hook).

As for custom indexing and an escape hatch for things prisma doesn't support -- I think those belong in the declarative `prisma.schema` rather than in migrations, but to be honest I haven't given that much thought to that since it's not something that I use in my day-to-day projects.

**Comment by cannikin** [comment]
We've got a PR for adding Data Migrations natively to RedwoodJS (here are the docs): https://deploy-preview-256--redwoodjs.netlify.app/docs/data-migrations 

It was inspired by the [data-migrate](https://github.com/ilyakatz/data-migrate) gem for Rails.

We now have three tasks that run during a deploy (`db save` is our wrapper around `prisma migrate up --experimental`):

    yarn redwood db up && yarn redwood dataMigrate up && yarn redwood build

It never occurred to me that Prisma could add this functionality natively so I just started building our own. 😬 But ideally these *would* tie into the migrate workflow so they can run in the proper order. Right now our process does have the downside that Prisma migrations and these data migrations are run completely separately, so you need to be careful if you're migrating data that could result in data loss, like removing a column. You need to deploy in two steps:

1. Add new table/column and migrate date
2. Remove old table/column

**Comment by AviBueno** [comment]
Have similar needs, one example (similar to @ryanking1809's) is detailed in https://github.com/prisma/prisma/discussions/3470
Looking forward to updates on the subject.

**Comment by Mautriz** [comment]
Same here, the possibility to use imperative sql alongside declarative migrations would solve all my problems, and I would use prisma for every project. As of now I'm kind of confused on how to do custom migrations in order. Would br awesome

**Comment by erawk** [comment]
This is a blocker for us. We're rolling our own post-deploy data migration scripts. Having something similar to Rails's `ActiveRecord` and `data-migrate` would be fantastic. In the mean time, we're doing something similar to Redwood above.

**Comment by AlexMenor** [comment]
I'm interested in this too! So I can declare views/indexes/on cascade statements and have my db ready with one command.

**Comment by albertoperdomo** [comment]
Hello,

I'm a product manager at Prisma, currently working on schema migrations.

As previously mentioned in this issue, we are working hard on taking Prisma Migrate to General Availability. As part of this work, we are making some important changes to how it works.
One aspect that's changing is that migrations will now be auto-generated but imperative, and there will be an option to create an empty migration file and implement your own logic inside of it.

We have an Early Access program for the Prisma Migrate with these set of changes and are looking for users who can test it and provide feedback. You can find more info and join the conversation in our [#product-feedback channel on our public Slack community](https://prisma.slack.com/archives/C01739JGFCM). You can join the Slack community [here](https://www.prisma.io/community).

**Comment by tomhoule** [comment]
As of 2.13.0, the new migrate with imperative migrations (as SQL, currently) in in preview. Feedback is very much wanted :)

**Comment by AlexMenor** [comment]
I'm using it already and I'll report as needed
